### PR TITLE
Add dynamic page title via router-view meta data

### LINKED ISSUE #343: Page title does not update whilst navigating the frontend
### Current Behavior

The page title of the forge app is always just `FlowForge` regardless of what page you are on.

This leads to a browser history that is impossible to navigate:

image

### Expected Behavior

The page title should include context of what page you are on

### Steps To Reproduce

 - Navigate around the front end
 - Look at the history

### Environment

- FlowForge version: 0.2.0

### PR DESCRIPTION
**Pull request by joepavitt**
Closes #343

### CODE REVIEW

File: frontend/src/routes.js [hunk 1]
```diff
@@ -34,4 +37,51 @@ const router = createRouter({
     routes
 })
 
+// This callback runs before every route change, including on page load.
+router.beforeEach((to, from, next) => {
+    console.log(from, to)
```
Comment (knolleary) [inline frontend/src/routes.js:42]:
```suggestion
```

File: frontend/src/routes.js [hunk 2]
```diff
@@ -34,4 +37,51 @@ const router = createRouter({
...
+    // e.g., if we have `/some/deep/nested/route` and `/some`, `/deep`, and `/nested` have titles,
+    // `/nested`'s will be chosen.
+    const nearestWithTitle = to.matched.slice().reverse().find(r => r.meta && r.meta.title)
+
+    // Find the nearest route element with meta tags.
+    const nearestWithMeta = to.matched.slice().reverse().find(r => r.meta && r.meta.metaTags)
+
+    const previousNearestWithMeta = from.matched.slice().reverse().find(r => r.meta && r.meta.metaTags)
+
+    console.log('title', nearestWithTitle)
```
Comment (knolleary) [inline frontend/src/routes.js:53]:
```suggestion
```

File: frontend/src/routes.js [hunk 3]
```diff
@@ -34,4 +37,51 @@ const router = createRouter({
...
+
+    // If a route with a title was found, set the document (page) title to that value.
+    if (nearestWithTitle) {
+        document.title = 'FlowForge | ' + nearestWithTitle.meta.title
+    } else if (previousNearestWithMeta) {
+        document.title = 'FlowForge | ' + previousNearestWithMeta.meta.title
+    }
```
[the remark this answers is not in the document]
Comment (joepavitt) [reply frontend/src/routes.js:58]: I wouldn't be against that. Happy to modify.

[the remark this answers is not in the document]
Comment (joepavitt) [reply frontend/src/routes.js:58]: Any preference on format?

Screenshot 2022-04-06 at 16 58 18

Could also do:

- `Team | Settings | FlowForge` - feels a bit strange re: hierarchy
- `Team | Settings - FlowForge` - not sure I like the `|` being a smaller break than `-`

[the remark this answers is not in the document]
Comment (knolleary) [reply frontend/src/routes.js:58]: `Team - Members - FlowForge` works for me

I think there is merit having the actual Team name (or Project name) in place of 'Team' and 'Project' - but lets go as above for now unless its a super quick and obvious switch out to include the team name/project name.

**Review by knolleary — APPROVED** [review]

**Pull request merged**